src/Common/src/System/Dynamic/Utils/ExpressionUtils.cs

```diff
@@ -226,5 +226,42 @@ internal static ParameterInfo[] GetParametersForValidation(MethodBase method, Ex
             }
             return pis;
         }
+
+        internal static bool SameElements<T>(IEnumerable<T> replacement, IReadOnlyList<T> current) where T : class
+        {
+            Debug.Assert(current != null);
+            Debug.Assert(replacement == null | replacement is ICollection<T>);
```

Comment:
I'm pretty sure I had a good reason for this strange thing, but can't think what it is right now. Will give it a look soon.